game-core/src/main/java/games/strategy/triplea/attachments/UnitAttachment.java

```diff
@@ -2031,6 +2031,10 @@ private void setFuelFlatCost(final IntegerMap<Resource> value) {
   }
 
   public IntegerMap<Resource> getFuelFlatCost() {
+    // TODO: remove for incompatible release
```

Comment:
@ron-murhammer does this assume we will update all maps to have fuel flat cost?  Why does an incompatible release update the situation?

Reply:
No. Both fuel cost models will be supported. This method is called for calculations and the expectation is that the value isn't null (which for all new games will be true). This is to avoid NPEs with save games since the member variable won't be initialized.

Reply:
Thanks for clarification.

> "This is to avoid NPEs with save games since the member variable won't be initialized."

Would that not continue to be the case well after an incompatible release? Save games have very long life spans, 6 months is not uncommon for a single save game.

Reply:
@DanVanAtta My definition of incompatible release is that it will no longer work with the old lobby or old save games. These would need finished using the old engine and old lobby. Otherwise code like this never goes away if we don't break save game compatibility at some point.

Reply:
Makes sense, thanks.

This is a really good example of why map compat is a problem (https://forums.triplea-game.org/topic/631/map-to-engine-compatibility-problems/28)

Thanks, that helps. I think I've the same understanding now.

This does I think clear note the compatibility problems we face. Fundamentally I think the key to it is that the lifespan of save games far outlives old lobby. I clearly remember how many players do not like having their save games killed, either because there is no lobby to play them in, or because everyone has an updated map that they were forced to update to so they could start new games. It used to be an ongoing problem that due to these issues some players on every release would abandon the game, the final straw.

It's really these problems that I'd like to get in to with the ongoing forum post: https://forums.triplea-game.org/topic/631/map-to-engine-compatibility-problems/28

Reply:
@DanVanAtta I think you are still misunderstanding. This has nothing to do with map versions or compatibility. It has to do with save games vs new games and the engine. So when a new game is started after this change, it initializes this new field to an empty collection so no chance of NPE. For save games started before this engine version, the field doesn't exist so when the save game is deserialized it sets the field to null making it so a NPE is possible.

So once every active save game out there is started with this engine version or later, there would be no chance of the NPE. Or if we decide to make a release that breaks save game incompatibility then we are purposefully choosing to not all save games to be loaded with the new engine version so this check could be removed.

Reply:
The proliferation of these `null` checks is making me wonder if we should add a `readObject()` method in such classes that initializes all the new fields to a meaningful value if they are `null` after deserialization.  That would consolidate all these compatibility checks in a single location per class and make it automatic during a review to check for a few new lines of code in the `readObject()` method if we see that a new field was added in a serializable type.

I think we talked about this in the past.  At that time, there didn't seem to be so many ("many" being relative, as I think there have been only about 10 such new fields added in the past year), and I think we felt that a compatibility-breaking release was a lot closer, and we would clean these up rather quickly. :smile:

Reply:
I like the suggestion for a 'readObject()' to do that work.

Reply:
Yeah, working towards an incompatible release has really dragged out much longer than I expected and want. It seems we keep introducing bugs to prevent cutting a new stable. Really trying to push to actually do it this time around :)